spouse/spouse_demo.py

```diff
@@ -0,0 +1,363 @@
+)
+
+# %%
+import matplotlib.pyplot as plt
+
+Y_probs_train = label_model.predict_proba(train_L)
+plt.hist(Y_probs_train[:, 1], bins=20, range=(0.0, 1.0))
+plt.show()
+
+# %% [markdown]
+# ### Part 4: Training our End Extraction Model
+#
+# In this final section of the tutorial, we'll use our noisy training labels alongside the development set labels to train our end machine learning model. We start by filtering out training examples which did not recieve a label from any LF, as these examples contain no signal. Then we concatenate them with dev set examples.
+#
+# %%
+from snorkel.analysis.utils import preds_to_probs
+from snorkel.labeling.utils import filter_unlabeled_dataframe
+
+# Change dev labels 1D array to 2D probabilities array as required for training end model.
+Y_probs_dev = preds_to_probs(Y_dev, 2)
+
+Y_probs_train = label_model.predict_proba(train_L)
+df_train_filtered, Y_probs_train_filtered = filter_unlabeled_dataframe(
+    X=df_train, y=Y_probs_train, L=train_L
+)
+
+df_combined = pd.concat([df_dev, df_train_filtered])
+Y_probs_combined = np.concatenate([Y_probs_dev, Y_probs_train_filtered], 0)
+
+# %% [markdown]
+# ### II. Training a _Long Short-term Memory_ (LSTM) Neural Network
+#
+# [LSTMs](https://en.wikipedia.org/wiki/Long_short-term_memory) can acheive state-of-the-art performance on many text classification tasks. We'll train a simple LSTM model below. tf_model contains functions for processing features and building the tensorflow graphs for training and evaluation.
+#
+# For this tutorial, we will be training a fairly effective deep learning model. More generally, however, Snorkel plugs in with many ML libraries, making it easy to use almost any state-of-the-art model as the end model!
+
+# %%
+from tf_model import get_model
+
+
+# Truncate sentences to limit memory usage when padding.
+def get_feature_arrays(df):
+    def pad_or_truncate(l, max_length=40):
+        return l[:max_length] + [""] * (max_length - len(l))
+
+    tokens = np.array(list(map(pad_or_truncate, df.tokens)))
+    idx1 = np.array(list(map(list, df.person1_word_idx)))
+    idx2 = np.array(list(map(list, df.person2_word_idx)))
+    return tokens, idx1, idx2
+
+
+model = get_model()
+tokens, idx1, idx2 = get_feature_arrays(df_combined)
+
+batch_size = 64
+num_epochs = 20  # TODO: Change this to ~50. Warning: Training takes several minutes!
+model.fit(
+    (tokens, idx1, idx2), Y_probs_combined, batch_size=batch_size, epochs=num_epochs
+)
+
+# %% [markdown]
+# Measure the trained model's prediction F1 score and ROC_AUC.
+
+# %%
+test_tokens, test_idx1, test_idx2 = get_feature_arrays(df_test)
+probs = model.predict((test_tokens, test_idx1, test_idx2))
+preds = probs_to_preds(probs)
+print(
+    f"Test F1 when trained with soft labels: {metric_score(Y_test, preds=preds, metric='f1')}"
+)
+print(
+    f"Test ROC-AUC when trained with soft labels: {metric_score(Y_test, probs=probs, metric='roc_auc')}"
+)
```

Comment:
Can we add a few conclusion sentences about what we've achieved, etc. so the takeaways/gains are crystal clear?